### ISSUE TITLE
[Feature] Run a request via script

### ISSUE THREAD

**Issue by mirkogolze** [usebruno/bruno#1350]
I think it was mentioned in the discussions around Bruno. But I create here a feature request (If duplicate please close).

It would be useful to have the ability to start a request via script.
E.g. to run a dependent pre-request.

The method in the bru-object should return the result of the assertions and the tests and give access to the response object as accessible in the post response script.

```javascript
  const result = await bru.runRequest('[folder]/[requestFile].bru');
```

**Comment by mj-h** [comment]
Would `bru.setNextRequest(<requestName>`) help you? It does not allow you to run a request from inside a request, but it allows you to customize the request chaining.

Note: Jumping back to earlier requests / the same request currently confuses the UI, but under the hood (and in the CLI) it runs fine.

**Comment by mirkogolze** [comment]
I want to use it in a pre-request-script. It think `bru.setNextRequest(...)` will not perform the chosen request to run before the current request. Could work as a workaround, but would lead to bad workaround request constellations.

**Comment by helloanoop** [comment]
`const result = await bru.runRequest('[folder]/[requestFile].bru');`

There have been multiple asks for this feature. Would love to get a PR for this.

**Comment by mj-h** [comment]
@helloanoop : I'm not sure about the "good first issue" tag -- the feature needs to be implemented in the CLI and in the UI. The UI tends to glitch when a request is being run multiple times.

**Comment by mj-h** [comment]
@helloanoop : Here is a video of a colllection, where I use `bru.setNextRequest()` to jump back from the third to the first request.

Two issues:
1. The UI appends new items to the list of requests that have been run, but updates the _state_ of the first instances
    - @bdentino attempted to fix this [here](https://github.com/mj-h/bruno/pull/1), but you had to [revert it](https://github.com/usebruno/bruno/issues/1215#issuecomment-1859306353)
2. There is no way to stop the infinite loop from the UI.

https://github.com/usebruno/bruno/assets/20513065/26a37aaa-89ae-4267-acd4-63cc7f8d4a34

**Comment by deepakthankachan** [comment]
@mj-h could you please do a code example of how you set the next request? thanks

**Comment by orached** [comment]
Hello, 

> @helloanoop : Here is a video of [a colllection](https://github.com/mj-h/setNextRequest-ui-bug-demo) that uses `bru.setNextRequest()` to jump back from the third to the first request.
> 
> Two issues:
> 
> 1. The UI appends new items to the list of requests that have been run, but updates the _state_ of the first instances
>    
>    * @bdentino attempted to fix this [here](https://github.com/mj-h/bruno/pull/1), but you had to [revert it](https://github.com/usebruno/bruno/issues/1215#issuecomment-1859306353)
> 2. There is no way to stop the infinite loop from the UI.
> 
>  Bildschirmaufnahme.2024-01-15.um.22.35.44.mov

Would this glitch e resolved?

**Comment by fayeulle** [comment]
Hello,

I'm very interested in this feature. I'd like to execute an "auth_request" within the current collection to populate "token" variables.

Is it possible to create a method within the Collection - Settings - Pre-request Script to achieve that?

```js
const result = await bru.runRequest('auth.bru');
```
Like this
![image](https://github.com/user-attachments/assets/4cde1e20-1276-4b58-9a17-74954618d9b9)

**Comment by gowriter** [comment]
you can use any http client library to do the work like axios, take a look here:
`const axios = require('axios')


const baseUrl = bru.getEnvVar("baseUrl")
const axiosInstance = axios.create({
  baseURL: baseUrl + "auth/",
  headers: {
  'Content-Type': 'application/x-www-form-urlencoded',
  },
});

const data = {
  "phone": bru.getEnvVar("authPhone"),
  "phonePrefix": "+34",
  "language": "es"
}

axiosInstance.post('requestPhoneValidation', data)
.catch(error => {
console.error('Authentication failed:', error);
});


const loginData = {
  "token": bru.getEnvVar("smsCode"),
  "id": bru.getEnvVar("authPhone"),
  "data": "+34",
  "type": "phone",
}

const response = await axiosInstance.post('login', loginData).catch(error => {
console.error('Authentication failed when login:', error);
});

const accessToken = response.data.tokens.token;
console.log(accessToken)
await bru.setVar("token", accessToken)

console.log("Token variable set")`

**Comment by fayeulle** [comment]
Yes @gowriter, but I wrote all logic in a dedicated request. 

It would be interesting to not rewrite request with axios and just use .bru files.
Authentication is just an example. I may have many actions to execute before each request

**Comment by mirkogolze** [comment]
Another very important thing is to benefit of the proxy configuration of Bruno. With scripting you do not benefit from the applications configuration and have to handle this manually again.

**Comment by deepakacharya** [comment]
agree with @thiagomini 
The error is being thrown when run from cli.
Error is something like this
`folder\request (bru.runRequest is not a function)`

**Comment by sreelakshmi-bruno** [comment]
Hi, we're working on the implementing these functions in the CLI. Will post updates here!

**Comment by ArturoBlazquez** [comment]
Hi, don't get autocomplete for this new method on Bruno 1.38.1
However it is working fine on the UI if I type the method manually.

(See that on the screenshot I get hints for `runner` but not for `runRequest`
![Screenshot 2025-01-27 at 15 56 33](https://github.com/user-attachments/assets/e274865c-aed1-4322-80a9-1c2abe9bb055)

**Comment by sreelakshmi-bruno** [comment]
Hi @ArturoBlazquez , we're working on the autocomplete. Will post updates here about the release.

**Comment by tomdracz** [comment]
We've implemented some re-request scripting to run a dependant request using `bru.runRequest` but it seems like the dependencies cannot cross collection boundaries.

In our case we had collection A and B and request in collection B trying to call request in collection A. Seemingly this is not supported and we've only managed to get around this by using a symlink to link collection A to Collection B/dependencies folder.

Is that the expected functionality? If so, any plans to support running requests from multiple collections?

**Comment by sreelakshmi-bruno** [comment]
Hi @tomdracz, this was intentional, assuming similar requests would be within the same collection. That said, I agree it would be a great enhancement to allow referencing requests from other collections. I'll update you on the implementation timeline. Thanks!

**Comment by ahujadevansh** [comment]
const result = await bru.runRequest('[folder]/[requestFile].bru');
not working properly in collection level pre request script but working in request level pre request script.

**Comment by anusree-bruno** [comment]
@ahujadevansh Yep, I can reproduce this! Since collection scripts run for every request, using bru.runRequest() creates an infinite loop (each request triggers the script again). We'll get this fixed soon 👍

**Comment by sreelakshmi-bruno** [comment]
Hi @ahujadevansh , I've created a separate issue for tracking purpose, see [here](https://github.com/usebruno/bruno/issues/4537). 

@tomdracz , due to technical limitations, it is not possible to reference request from another collection. Although we could add the functionality to reference requests from a different folder when executing them via folder run. Would grouping similar requests into folders, executing them via folder run work for your usecase? If so, please raise an enhancement request describing your usecase, Thanks!

Closing this issue since the bru.runRequest() is now available in the app (with autocomplete suggestion) and the CLI. The follow up issues are being tracked in separate issues as mentioned. Thanks!
